[ONNX] Align Optional Type in block

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83186
* #83201
* __->__ #83599

Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

titaiwangms committed:
[ONNX] Align Optional Type in block

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"




[ghstack-poisoned]

AllenTiTaiWang commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -12233,7 +12233,7 @@ def forward(self, x) -> Optional[Tensor]:
 
     @common_utils.parametrize(
         "module_class",
-        (IfNoneOutput, IfNoneInput, LoopNoneOutput),
```
Reenable this test as it was blocked by this bug.

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -206,7 +206,9 @@ void ReplaceBlockOutputWithOptional(
   Node* opt_node = ONNXOptionalNode(opt_type, block->owningGraph());
```
not this line, but the comment above for this function needs update. It states "Needed when control flow has multiple branches ..." which does not apply to "loop"

AllenTiTaiWang replied:
As if and loop share this function, I keep the original description and add more comment of loop for this function.

BowenBao commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -206,7 +206,9 @@ void ReplaceBlockOutputWithOptional(
   opt_node->insertBefore(block->return_node());
   Value* block_output = block->outputs().at(i);
-  block_output->replaceAllUsesWith(opt_node->output());
+  // replace only the last node as Optional Node only affects
```
Suggest rephrasing, since the code is not replacing the node, but replacing the value.

AllenTiTaiWang replied:
Done

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

AllenTiTaiWang commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -930,6 +930,14 @@ void Value::replaceFirstUseWith(Value* newValue) {
 }
 
+void Value::replaceLastUseWith(Value* newValue) {
+  AT_ASSERT(owningGraph() == newValue->owningGraph());
```
Done. I saw a lot of other functions still using AT_ASSERT. Maybe an issue should be raised and clean all those?

AllenTiTaiWang replied:
https://github.com/pytorch/pytorch/issues/84132

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

BowenBao reviewed (approved):
LGTM, CI seems unhappy, might need a rebase for the stack.

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

justinchuby commented:
Q: what does "aligning" mean in the title?

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g

AllenTiTaiWang commented:
@malfet Could you take a look? It's a bug fix on ONNX Optional type in Loop.

malfet commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -251,6 +251,7 @@ struct Value {
   }
 
   TORCH_API void replaceFirstUseWith(Value* newValue);
+  TORCH_API void replaceLastUseWith(Value* newValue);
```
Why this and firstUse needs to be a `TORCH_API`? I.e. it it supposed to be called from the external library (correct me if I'm wrong, but both torch/csrc/jit/ir/ir.cpp and torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp are compiled as part of `libtorch_cpu`?)

The reason I'm asking this: Windows have hardcoded max of 64K exported entires and TorchCPU exposes 60+k of them right now.

AllenTiTaiWang replied:
A revisit on replacing nodes API that we already had, I think in this case, we can utilize on an exisiting one. For TORCH_API, taking out TORCH_API from `replaceFirstUseWith` breaks libtorch. I guess cause `replaceFirstUseWith` is legacy and used in `replaceAllUseWith` which is heavily used across whole Pytorch repo, so I will remain that one.

malfet reviewed (approved):
LGTM

AllenTiTaiWang commented:
> Q: what does "aligning" mean in the title?

In Loop, we had Optional Type as input without having Optional Type as output.

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g

titaiwangms committed:
Update on "[ONNX] Align Optional Type in block"


Why:

Previously, we use `replaceAlluseswith` after adding Optional on the node which is right before output. However, this may break the graph by also changing the nodes that needs the node (original) as input. We only need the node to be optional in output.

[ghstack-poisoned]

AllenTiTaiWang commented:
@pytorchbot merge -g
